src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorAdornmentManager.cs

```diff
@@ -0,0 +1,216 @@
+                    if (point == null)
+                    {
+                        continue;
+                    }
+
+                    var mappedPoint = TextView.BufferGraph.MapUpToSnapshot(
+                        point.Value, PointTrackingMode.Negative, PositionAffinity.Predecessor, TextView.VisualSnapshot);
+                    if (mappedPoint == null)
+                    {
+                        continue;
+                    }
+
+                    if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, viewSnapshot, out var span))
+                    {
+                        continue;
+                    }
+
+                    if (!viewLines.IntersectsBufferSpan(span))
+                    {
+                        // span is outside of the view so we will not get geometry for it, but may 
+                        // spent a lot of time trying.
+                        continue;
+                    }
+
+                    var lineNum = mappedPoint.Value.GetContainingLine().LineNumber;
+                    if (!map.TryGetValue(lineNum, out var list))
+                    {
+                        list = new List<IMappingTagSpan<InlineErrorTag>>();
+                        map.Add(lineNum, list);
+                    }
+
+                    list.Add(tagMappingSpan);
+                    _tagSpanToPointMap.Add(tagMappingSpan, point.Value);
+                }
+            }
+
+            return map.ToImmutable();
+        }
+
+        protected override void UpdateSpans_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection, bool removeOldTags)
+        {
+            Contract.ThrowIfNull(changedSpanCollection);
+
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            if (viewLines == null || viewLines.Count == 0)
+            {
+                return; // nothing to draw on
+            }
+
+            // removing is a separate pass from adding so that new stuff is not removed.
+            if (removeOldTags)
+            {
+                foreach (var changedSpan in changedSpanCollection)
+                {
+                    // is there any effect on the view?
+                    if (viewLines.IntersectsBufferSpan(changedSpan))
+                    {
+                        AdornmentLayer.RemoveAdornmentsByVisualSpan(changedSpan);
+                    }
+                }
+            }
+
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            foreach (var (lineNum, tagMappingSpanList) in map)
+            {
+                var tagMappingSpan = GetHighestOrderTag(tagMappingSpanList);
+                if (tagMappingSpan != null)
+                {
+                    TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span);
```

Comment:
no check if this fails?

Reply:
should never fail since it's checked in the "GetSpansOnEachLine", I needed to call it again to retrieve the span. There is probably a better way to do this to not call it twice.